File: ee/query-service/model/pipelines.go
package model

import (
	"encoding/json"

	"github.com/pkg/errors"
)

// Pipeline is stored and also deployed finally to collector config
type Pipeline struct {
	Id      string `json:"id,omitempty" db:"id"`
	OrderId string `json:"orderId" db:"order_id"`
	Name    string `json:"name,omitempty" db:"name"`
	Alias   string `json:"alias" db:"alias"`
	Enabled bool   `json:"enabled" db:"enabled"`
	Filter  string `json:"filter" db:"filter"`
	// configuration for pipeline
	RawConfig string `db:"config_json"`

	Config []PipelineOperator `json:"config"`

	Creator
	Updater
}

type Processor struct {
	Operators []PipelineOperator `json:"operators" yaml:"operators"`
}

type PipelineOperator struct {
	Type        string           `json:"type" yaml:"type" mapstructure:"type"`
	ParseTo     string           `json:"parse_to,omitempty" yaml:"parse_to,omitempty" mapstructure:"parse_to"`
	Pattern     string           `json:"pattern,omitempty" yaml:"pattern,omitempty" mapstructure:"pattern"`
	Output      string           `json:"output,omitempty" yaml:"output,omitempty" mapstructure:"output"`
	Regex       string           `json:"regex,omitempty" yaml:"regex,omitempty" mapstructure:"regex"`
	ID          string           `json:"id,omitempty" yaml:"id,omitempty" mapstructure:"id"`
	ParseFrom   string           `json:"parse_from,omitempty" yaml:"parse_from,omitempty" mapstructure:"parse_from"`
	Timestamp   *TimestampParser `json:"timestamp,omitempty" yaml:"timestamp,omitempty" mapstructure:"timestamp"`
	TraceParser *TraceParser     `json:"trace_parser,omitempty" yaml:"trace_parser,omitempty" mapstructure:"trace_parser"`
	Field       string           `json:"field,omitempty" yaml:"field,omitempty" mapstructure:"field"`
	Value       string           `json:"value,omitempty" yaml:"value,omitempty" mapstructure:"value"`
	From        string           `json:"from,omitempty" yaml:"from,omitempty" mapstructure:"from"`
	To          string           `json:"to,omitempty"  yaml:"to,omitempty" mapstructure:"to"`
	Expr        string           `json:"expr,omitempty" yaml:"expr,omitempty" mapstructure:"expr"`
	Routes      *[]Route         `json:"routes,omitempty" yaml:"routes,omitempty"`
	Fields      []string         `json:"fields,omitempty" yaml:"fields,omitempty"`
	Default     string           `json:"default,omitempty" yaml:"default,omitempty"`
}

type TimestampParser struct {
	Layout     string `json:"layout" yaml:"layout"`
	LayoutType string `json:"layout_type" yaml:"layout_type"`
	ParseFrom  string `json:"parse_from" yaml:"parse_from"`
}

type TraceParser struct {
	TraceId    *ParseFrom `json:"trace_id,omitempty" yaml:"trace_id,omitempty"`
	SpanId     *ParseFrom `json:"span_id,omitempty" yaml:"span_id,omitempty"`
	TraceFlags *ParseFrom `json:"trace_flags,omitempty" yaml:"trace_flags,omitempty"`
}

type ParseFrom struct {
	ParseFrom string `json:"parse_from" yaml:"parse_from"`
}

type Route struct {
	Output string `json:"output" yaml:"output"`
	Expr   string `json:"expr" yaml:"expr"`
}

func (i *Pipeline) ParseRawConfig() error {
	c := []PipelineOperator{}
	err := json.Unmarshal([]byte(i.RawConfig), &c)
	if err != nil {
		return errors.Wrap(err, "failed to parse ingestion rule config")
	}
	i.Config = c
	return nil
}

File: frontend/src/pages/LogsModulePage/Pipelines/stories/__story_mockdata__/pipelines.ts
/**
 * AI-owned. Generated and maintained by the `signoz-page-story` skill.
 * Do not hand-edit: regenerate instead.
 */

import type { ILog } from 'types/api/logs/log';
import type {
	HistoryData,
	Pipeline,
	PipelineData,
	ProcessorData,
} from 'types/api/pipeline/def';
import { DataTypes } from 'types/api/queryBuilder/queryAutocompleteResponse';
import type { TagFilter } from 'types/api/queryBuilder/queryBuilderData';

/**
 * What the API sends. The change-history column renders the lowercase spelling,
 * so a version's stage stays blank there: see the story's PR.
 */
export const DEPLOY_STATUSES = [
	'DEPLOYED',
	'IN_PROGRESS',
	'FAILED',
	'DIRTY',
	'UNKNOWN',
] as const;

export type DeployStatus = (typeof DEPLOY_STATUSES)[number];

const DEPLOY_RESULTS: Record<DeployStatus, string> = {
	DEPLOYED: 'deploy success',
	IN_PROGRESS: 'deployment started',
	FAILED: 'collector rejected the config: unknown processor "grok_parser"',
	DIRTY: 'config on the collector is older than this version',
	UNKNOWN: '',
};

const filterOn = (key: string, value: string): TagFilter => ({
	op: 'AND',
	items: [
		{
			id: `${key}-${value}`,
			key: { key, dataType: DataTypes.String, type: '' },
			op: '=',
			value,
		},
	],
});

type ProcessorSeed = Omit<ProcessorData, 'orderId' | 'id' | 'output'>;

const PROCESSORS: ProcessorSeed[] = [
	{
		type: 'grok_parser',
		name: 'Parse the access log line',
		enabled: true,
		pattern: '%{COMMONAPACHELOG}',
		parse_from: 'body',
		parse_to: 'attributes',
		on_error: 'send',
	},
	{
		type: 'json_parser',
		name: 'Parse the JSON payload',
		enabled: true,
		parse_from: 'body',
		parse_to: 'attributes',
		enable_flattening: true,
		enable_paths: false,
		path_prefix: '',
		mapping: {},
	},
	{
		type: 'severity_parser',
		name: 'Read severity off level',
		enabled: true,
		parse_from: 'attributes.level',
	},
	{
		type: 'move',
		name: 'Rename auth to username',
		enabled: true,
		from: 'attributes.auth',
		to: 'attributes.username',
	},
	{
		type: 'remove',
		name: 'Drop the raw payload',
		enabled: false,
		field: 'attributes.payload',
	},
];

export const PROCESSOR_MAX = PROCESSORS.length;

/**
 * `output` chains each processor to the next one, which is what cancelling an
 * edit rebuilds the list from.
 */
const processorsFor = (pipelineId: string, count: number): ProcessorData[] => {
	const ids = PROCESSORS.slice(0, count).map(
		(_unused, index) => `${pipelineId}-processor-${index + 1}`,
	);

	return PROCESSORS.slice(0, count).map((processor, index) => ({
		...processor,
		id: ids[index],
		orderId: index + 1,
		output: ids[index + 1],
	}));
};

interface PipelineSeed {
	name: string;
	description: string;
	filter: TagFilter;
}

const PIPELINES: PipelineSeed[] = [
	{
		name: 'Parse nginx access logs',
		description: 'grok the access log line into attributes',
		filter: filterOn('source', 'nginx'),
	},
	{
		name: 'Drop health check noise',
		description: 'filter out the /healthz requests',
		filter: filterOn('http.route', '/healthz'),
	},
	{
		name: 'Extract trace context',
		description: 'read trace_id and span_id from the body',
		filter: filterOn('service.name', 'checkout'),
	},
	{
		name: 'Rename severity field',
		description: 'level becomes severity_text',
		filter: filterOn('service.name', 'payments'),
	},
	{
		name: 'Flatten JSON payloads',
		description: 'expand nested attributes into their own keys',
		filter: filterOn('deployment.environment', 'production'),
	},
];

export const PIPELINE_MAX = PIPELINES.length;

export const VERSION_MAX = 6;

export interface PipelinesOptions {
	pipelines: number;
	processors: number;
	deployStatus: DeployStatus;
	versions: number;
}

const historyFor = (
	versions: number,
	deployStatus: DeployStatus,
): HistoryData[] =>
	Array.from({ length: versions }, (_unused, index) => {
		const version = versions - index;
		const isLatest = index === 0;
		const status = isLatest ? deployStatus : 'DEPLOYED';

		return {
			active: isLatest,
			createdAt: `2026-07-${String(10 + version).padStart(2, '0')}T13:48:31.032106578Z`,
			createdBy: isLatest ? 'ada@signoz.io' : 'grace@signoz.io',
			createdByName: isLatest ? 'Ada Lovelace' : 'Grace Hopper',
			deployStatus: status,
			deployResult: DEPLOY_RESULTS[status],
			disabled: false,
			elementType: 'log_pipelines',
			id: `storybook-pipelines-v${version}`,
			isValid: true,
			lastConf: '',
			lastHash: '',
			version,
		};
	});

/**
 * A latest version still deploying makes the page poll every three seconds,
 * which is the app's own behaviour and what the `Deploy status` control shows.
 */
export const logsPipelinesResponse = ({
	pipelines,
	processors,
	deployStatus,
	versions,
}: PipelinesOptions): { status: string; data: Pipeline } => ({
	status: 'success',
	data: {
		active: true,
		createdBy: 'ada@signoz.io',
		deployResult: DEPLOY_RESULTS[deployStatus],
		deployStatus,
		disabled: false,
		elementType: 'log_pipelines',
		id: 'storybook-pipelines',
		is_valid: true,
		lastConf: '',
		lastHash: '',
		version: versions,
		history: historyFor(versions, deployStatus),
		pipelines: PIPELINES.slice(0, pipelines).map(
			(pipeline, index): PipelineData => {
				const id = `storybook-pipeline-${index + 1}`;

				return {
					...pipeline,
					id,
					alias: pipeline.name.toLowerCase().replace(/\s+/g, '-'),
					createdAt: '2026-07-16T13:48:31.032106578Z',
					createdBy: 'ada@signoz.io',
					enabled: index !== 3,
					orderId: index + 1,
					config: processorsFor(id, processors),
				};
			},
		),
	},
});

/**
 * The simulation answers on the logs it was sent, so the preview can diff input
 * against output: the parsed fields land on the ones the pipeline writes to.
 */
export const pipelinePreviewResponse = (
	logs: ILog[],
): { status: string; data: { logs: ILog[] } } => ({
	status: 'success',
	data: {
		logs: logs.map((log) => ({
			...log,
			// `ILog` types every attribute map as `Record<string, never>`, so a value
			// only goes in through a cast.
			attributes_string: {
				...log.attributes_string,
				'http.method': 'POST',
				username: 'ada',
				level: String(log.severity_text ?? '').toLowerCase(),
			} as unknown as ILog['attributes_string'],
		})),
	},
});
